src/main/java/org/dataone/hashstore/hashstoreconverter/HashStoreConverter.java

```diff
@@ -0,0 +1,122 @@
+    private final FileHashStoreLinks fileHashStoreLinks;
+
+    /**
+     * Constructor to initialize HashStoreConverter. Properties to an existing or desired HashStore
+     * are required.
+     *
+     * @param hashstoreProperties Properties object with the following keys: storePath, storeDepth,
+     *                            storeWidth, storeAlgorithm, storeMetadataNamespace
+     * @throws IOException              Issue with directories or hashstore.yaml config
+     * @throws NoSuchAlgorithmException Unsupported FileHashStoreLinks algorithm
+     */
+    public HashStoreConverter(Properties hashstoreProperties)
+        throws IOException, NoSuchAlgorithmException {
+        try {
+            fileHashStoreLinks = new FileHashStoreLinks(hashstoreProperties);
+            logHashStoreConverter.info("HashStoreConverter initialized");
+
+        } catch (IOException ioe) {
+            String errMsg = "Unexpected IOException encountered: " + ioe.getMessage();
+            logHashStoreConverter.error(errMsg);
+            throw ioe;
+
+        } catch (NoSuchAlgorithmException nsae) {
+            String errMsg = "A supplied algorithm is not supported: " + nsae.getMessage();
+            logHashStoreConverter.error(errMsg);
+            throw nsae;
+
+        }
+    }
+
+    /**
+     * Take an existing path to a data object, store it into a new or existing HashStore via a
+     * hard link, store the supplied system metadata and return the ObjectMetadata for the data
+     * object. The 'sysMetaStream' given may never be null. A 'filePath' may be null, in which
+     * case a data object will not be stored.
+     *
+     * @param filePath      Path to existing data object
+     * @param pid           Persistent or authority-based identifier
+     * @param sysmetaStream Stream to sysmeta content to store.
+     * @param checksum   Value of checksum
+     * @param checksumAlgorithm Ex. "SHA-256"
+     * @return ObjectMetadata for the given pid
+     * @throws IOException              An issue with calculating checksums or storing sysmeta
+     * @throws NoSuchAlgorithmException An algorithm defined is not supported
+     * @throws InterruptedException     Issue with synchronizing storing metadata
+     */
+    public ObjectMetadata convert(Path filePath, String pid, InputStream sysmetaStream,
+                                  String checksum, String checksumAlgorithm)
+        throws IOException, NoSuchAlgorithmException, InterruptedException {
+        logHashStoreConverter.info("Begin converting data object and sysmeta for pid: " + pid);
+        FileHashStoreUtility.ensureNotNull(sysmetaStream, "sysmetaStream", "convert");
+        FileHashStoreUtility.ensureNotNull(pid, "pid", "convert");
+        FileHashStoreUtility.checkForEmptyAndValidString(pid, "pid", "convert");
+
+        // Store the sysmeta first - this can never be null and is always required.
+        try {
+            fileHashStoreLinks.storeMetadata(sysmetaStream, pid);
```

Comment:
I am thinking the order of creating. Which should be first, system metadata or hard links? I prefer to hard links since this is the main purpose. Failure of storing system metadata may disrupt the creation of the hard link, which may succeed. What do you think? If we create hard link first, I would like the method storeMetadata will throw a new exception, which can tell the failure happens in the storeMetadata method and indicate the hard link was created successfully.

Reply:
When I initially reversed the order, I had overlooked the hard linking process being the main purpose. After reviewing your comment, I agree with your suggestion.

With the current order, if `storeMetadata` fails, an exception will be thrown. So we will never reach the code to create a hard link - and won't be able to tell if there is another latent bug or issue affecting storing a hard link.

I have reverted my change, thank you!